File: it-agents/skills/code-architect/references/code-examples/csharp/TemplateMethod.cs
// Template Method — define an algorithm skeleton in a base class with overridable steps.
using System;

namespace Patterns.TemplateMethod
{
    abstract class AbstractProcessor
    {
        // The template method defines the fixed algorithm structure.
        public string Process(string input)
        {
            string read = ReadStep(input);
            string transformed = TransformStep(read);
            return WriteStep(transformed);
        }

        protected virtual string ReadStep(string input) => $"read({input})";
        protected abstract string TransformStep(string data); // required hook
        protected virtual string WriteStep(string data) => $"write({data})";
    }

    sealed class UpperProcessor : AbstractProcessor
    {
        protected override string TransformStep(string data) => data.ToUpperInvariant();
    }

    sealed class ReverseProcessor : AbstractProcessor
    {
        protected override string TransformStep(string data)
        {
            char[] chars = data.ToCharArray();
            Array.Reverse(chars);
            return new string(chars);
        }

        // Overrides an optional step.
        protected override string WriteStep(string data) => $"out<<{data}>>";
    }

    static class Demo
    {
        public static void Run()
        {
            Console.WriteLine(new UpperProcessor().Process("abc"));
            Console.WriteLine(new ReverseProcessor().Process("abc"));
        }

        public static void Main() => Run();
    }
}

File: it-agents/skills/code-architect/references/code-examples/cpp/TemplateMethod.cpp
// Template Method — algorithm skeleton in a base class with overridable steps.
#include <iostream>
#include <string>

// Base class defines the invariant skeleton and abstract steps.
class Importer {
public:
    virtual ~Importer() = default;

    // The template method: fixed sequence, customizable steps.
    void process(const std::string& source) {
        auto raw = read(source);
        auto parsed = parse(raw);
        store(parsed);
        hook();
    }

protected:
    virtual std::string read(const std::string& source) {
        return "<bytes of " + source + ">";
    }
    virtual std::string parse(const std::string& raw) = 0; // required step
    virtual void store(const std::string& parsed) {
        std::cout << "stored: " << parsed << "\n";
    }
    virtual void hook() {} // optional step with default no-op
};

class CsvImporter : public Importer {
protected:
    std::string parse(const std::string& raw) override { return "csv{" + raw + "}"; }
};

class JsonImporter : public Importer {
protected:
    std::string parse(const std::string& raw) override { return "json{" + raw + "}"; }
    void hook() override { std::cout << "json import audited\n"; }
};

int main() {
    CsvImporter{}.process("table.csv");
    JsonImporter{}.process("payload.json");
    return 0;
}
